bin/utest/pipe.c

```diff
@@ -104,3 +109,226 @@ int test_pipe_child_signaled(void) {
 
   return 0;
 }
+
+int test_pipe_blocking_flag_manipulation(void) {
+  int pipe_fd[2];
+
+  /* creating pipe */
+  int pipe2_ret = pipe2(pipe_fd, O_NONBLOCK);
+  assert(pipe2_ret == 0);
+
+  /* check if flag is set */
+  assert(fcntl(pipe_fd[0], F_GETFL) & O_NONBLOCK);
+  assert(fcntl(pipe_fd[1], F_GETFL) & O_NONBLOCK);
+
+  /* unset same flag for read end */
+  int read_flagset_with_block = fcntl(pipe_fd[0], F_GETFL);
+  assert(read_flagset_with_block > 0);
+  fcntl(pipe_fd[0], F_SETFL, read_flagset_with_block & ~O_NONBLOCK);
+  /* unset same flag for write end */
+  int write_flagset_with_block = fcntl(pipe_fd[1], F_GETFL);
+  assert(write_flagset_with_block > 0);
+  fcntl(pipe_fd[1], F_SETFL, write_flagset_with_block & ~O_NONBLOCK);
+
+  /* check if flag is not set */
+  assert(!(fcntl(pipe_fd[0], F_GETFL) & O_NONBLOCK));
+  assert(!(fcntl(pipe_fd[1], F_GETFL) & O_NONBLOCK));
+
+  close(pipe_fd[0]);
+  close(pipe_fd[1]);
+
+  return 0;
+}
+
+int test_pipe_write_interruptible_sleep(void) {
+  int pipe_fd[2];
+  pid_t child_pid;
+
+  /* creating pipe */
+  int pipe2_ret = pipe2(pipe_fd, 0);
+  assert(pipe2_ret == 0);
+
+  /* forking */
+  child_pid = fork();
+  assert(child_pid >= 0);
+
+  if (child_pid == 0) { /* child */
+    close(pipe_fd[0]);  /* closing read end of pipe */
+
+    struct sigaction sa = {
+      .sa_handler = handler,
+      .sa_flags = 0,
+    };
+
+    sigaction(SIGALRM, &sa, NULL);
+
+    int page_size = getpagesize();
+    char data[page_size];
```

Comment:
VLA is a horrible idea.

Reply:
What does "VLA" mean?

Reply:
variable-length array

Reply:
What do you suggest then? Set it to a constant value? Do you suggest any? Maybe Linux page size...?

Reply:
I prefer to allocate array on heap (if you don't know size). In our case page size is always 4096 - see implementation of `getpagesize` - so you can make it static or allocate whole array on stack.

Reply:
Yup. Please avoid VLA. IIRC It's not written down anywhere, but it's forbidden inside the kernel. We don't want to repeat mistakes Linux went through at some point.